sql-plugin/src/main/scala/org/apache/spark/sql/rapids/mathExpressions.scala

```diff
@@ -453,13 +453,160 @@ abstract class GpuRoundBase(child: Expression, scale: Expression) extends GpuBin
+    // and the -1e19 overflows as 8446744073709551616L. The overflow happens in the same way for
+    // HALF_UP (round) and HALF_EVEN (bround).
+    def fixUpInt64OnBounds(zeroFn: () => Scalar): ColumnVector = {
+      // Builds predicates on whether there is a round up on the max digit or not
+      val litForCmp = Seq(Scalar.fromLong(1000000000000000000L),
+                          Scalar.fromLong(4L),
+                          Scalar.fromLong(-4L))
+      val (needRep, needNegRep) = withResource(litForCmp) { case Seq(base, four, minusFour) =>
+        withResource(lhsValue.div(base)) { headDigit =>
+          closeOnExcept(headDigit.greaterThan(four)) { posRep =>
+            closeOnExcept(headDigit.lessThan(minusFour)) { negRep =>
+              posRep -> negRep
+            }
+          }
+        }
+      }
+      // Replaces with corresponding literals
+      val litForRep = Seq(zeroFn(),
+                          Scalar.fromLong(8446744073709551616L),
+                          Scalar.fromLong(-8446744073709551616L))
+      val repVal = withResource(litForRep) { case Seq(zero, upLit, negUpLit) =>
+        withResource(needRep) { _ =>
+          withResource(needNegRep) { _ =>
+            withResource(needNegRep.ifElse(upLit, zero)) { negBranch =>
+              needRep.ifElse(negUpLit, negBranch)
+            }
+          }
+        }
+      }
+      // Handles null values
+      withResource(repVal) { _ =>
+        if (lhsValue.hasNulls) {
+          repVal.mergeAndSetValidity(BinaryOp.BITWISE_AND, lhsValue)
+        } else {
+          repVal.incRefCount()
+        }
+      }
+    }
+
+    // Fixes up float points rounded by a scale exceeding the max digits of data type. Under this
+    // circumstance, cuDF produces different results to Spark.
+    // Compared to integral values, fixing up round overflow of float points needs to take care
+    // of some special values: nan, inf, -inf.
+    def fpZeroReplacement(zeroFn: () => Scalar,
+        infFn: () => Scalar,
+        negInfFn: () => Scalar): ColumnVector = {
+      val scaleVal = scale.getValue.asInstanceOf[Int]
+      val maxDigits = if (dataType == FloatType) 39 else 309
+      if (-scaleVal >= maxDigits) {
+        // replaces common values (!Null AND !Nan AND !Inf And !-Inf) with zero, while keeps
+        // all the special values unchanged
+        withResource(Seq(zeroFn(), infFn(), negInfFn())) { case Seq(zero, inf, negInf) =>
+          // builds joined predicate: !Null AND !Nan AND !Inf And !-Inf
+          val joinedPredicate = {
+            val conditions = Seq(() => lhsValue.isNotNan,
+                                 () => lhsValue.notEqualTo(inf),
+                                 () => lhsValue.notEqualTo(negInf))
+            conditions.foldLeft(lhsValue.isNotNull) { case (buffer, builder) =>
+              withResource(buffer) { _ =>
+                withResource(builder()) { predicate =>
+                  buffer.and(predicate)
```

Comment:
Where are the intermediate values closed?

Reply:
If I am not wrong, the output of Nth foldLeft iteration will be closed in N+1th iteration as buffer.